### PR TITLE
fix web3 connection sync routine if nethermind:

### PR DESCRIPTION
**Pull request by jordipainan**
Adapt the sync progress routine to work properly
when the connected web3 endpoint runs the Nethermind
codebase. The eth_syncing JSON RPC call returns 0x0
as block number.

Add an extra call (eth_blockNumber) in the Connect
method to check that the client can obtain at least the
block number from the connected web3 endpoint.

### CODE REVIEW

File: ethereum/handler/ethereumHandler.go [hunk 1]
```diff
@@ -134,8 +142,22 @@ func (eh *EthereumHandler) SyncInfo(ctx context.Context) (*EthSyncInfo, error) {
...
 	if sp != nil {
-		info.MaxHeight = sp.HighestBlock
-		info.Height = sp.CurrentBlock
+		// some clients return 0x0 on block number if are synced
+		// do an extra check getting the latest block number via eth_blockNumber
+		ctx, cancel := context.WithTimeout(context.Background(), time.Second*30)
+		defer cancel()
+		blkNum, err := eh.EthereumClient.BlockNumber(ctx)
+		if err != nil {
+			log.Debugf("cannot connect: %s", err)
```
Comment (p4u) [inline ethereum/handler/ethereumHandler.go]: if the error is returned it should not need to be printed

File: ethereum/handler/ethereumHandler.go [hunk 2]
```diff
@@ -134,8 +142,22 @@ func (eh *EthereumHandler) SyncInfo(ctx context.Context) (*EthSyncInfo, error) {
 		return nil, err
 	}
 	if sp != nil {
-		info.MaxHeight = sp.HighestBlock
-		info.Height = sp.CurrentBlock
+		// some clients return 0x0 on block number if are synced
+		// do an extra check getting the latest block number via eth_blockNumber
+		ctx, cancel := context.WithTimeout(context.Background(), time.Second*30)
+		defer cancel()
+		blkNum, err := eh.EthereumClient.BlockNumber(ctx)
```
Comment (p4u) [inline ethereum/handler/ethereumHandler.go]: use info.Height directly

File: ethereum/handler/ethereumHandler.go [hunk 3]
```diff
@@ -134,8 +142,22 @@ func (eh *EthereumHandler) SyncInfo(ctx context.Context) (*EthSyncInfo, error) {
 		return nil, err
 	}
 	if sp != nil {
```
Comment (p4u) [inline ethereum/handler/ethereumHandler.go]: lets reverse this
```
if sp == nil {
	header, err := eh.EthereumClient.HeaderByNumber(ctx, nil)
		if err != nil {
			return nil, err
		}
		info.Height = uint64(header.Number.Int64())
		info.MaxHeight = info.Height
		info.Synced = info.Height > 0
		return info,nil
		}
```

**Review by p4u — APPROVED** [review]

**Comment by p4u** [comment]
We should cherry-pick this commit to production

File: ethereum/handler/ethereumHandler.go [hunk 4]
```diff
@@ -133,17 +141,28 @@ func (eh *EthereumHandler) SyncInfo(ctx context.Context) (*EthSyncInfo, error) {
...
-	} else {
+	if sp == nil {
 		header, err := eh.EthereumClient.HeaderByNumber(ctx, nil)
 		if err != nil {
 			return nil, err
 		}
 		info.Height = uint64(header.Number.Int64())
 		info.MaxHeight = info.Height
 		info.Synced = info.Height > 0
+	} else {
```
Comment (p4u) [inline ethereum/handler/ethereumHandler.go:163]: The objective of the if reverse was to remove the "else" and add a return inside the if

Comment (jordipainan) [reply ethereum/handler/ethereumHandler.go:163]: done

**Comment by jordipainan** [comment]
@mvdan it is kind of strange that the staticcheck returns an error in this case:
`rpcapi/votehandlers.go:287:18: func (*RPCAPI).getProcessRollingCensusSize is unused (U1000)` while this commit does not introduce such change.
Do you have any ideas ?

**Comment by mvdan** [comment]
Assuming master is green, have you tried rebasing on the latest master?

**Comment by jordipainan** [comment]
@mvdan  force pushed the rebase but b68e96d58c571a2006c6ab85c0003f02d0033545 was the latest commit on this branch and is also green in master, now compose-test error :)

**Comment by p4u** [comment]
> @mvdan it is kind of strange that the staticcheck returns an error in this case: `rpcapi/votehandlers.go:287:18: func (*RPCAPI).getProcessRollingCensusSize is unused (U1000)` while this commit does not introduce such change. Do you have any ideas ?

I also saw this error in the past. It appears randomly. This comes from a lint ignore line added by @ed255 (but to be honest I thought I removed that line...)

EDIT: I misunderstood the error... nothing to do with what I found in the past

**Comment by jordipainan** [comment]
> > @mvdan it is kind of strange that the staticcheck returns an error in this case: `rpcapi/votehandlers.go:287:18: func (*RPCAPI).getProcessRollingCensusSize is unused (U1000)` while this commit does not introduce such change. Do you have any ideas ?
> 
> I also saw this error in the past. It appears randomly. This comes from a lint ignore line added by @ed255 (but to be honest I thought I removed that line...)

thank you for the input @p4u lets see what is happening with the tests now that I just did the rebase with the latest master commit

**Pull request merged**